### ISSUE TITLE
Improve GitHub labels

### ISSUE THREAD

**Issue by per1234** [107-systems/107-Arduino-NMEA-Parser#23]
Reference: https://github.com/107-systems/107-Arduino-NMEA-Parser/issues/20#issuecomment-719932632
>@per1234 can you replace the default GitHub labels with others which are a bit more useful? I'd need at least a "Bug Report" and a "Feature Request" label, if you know other labels that make sense then please add them too.

I have a rough idea about a standardized format for labels, as well as how to propagate them across all the repositories. This is a copy/paste from some of my WIP notes, so forgive me if they are not very polished.

Having inconsistent labels the repositories of an organization can make the maintainer's job more difficult as well as making the issue/PR trackers less useful to the users. It can be difficult to interpret the specific meaning of labels and when it is appropriate to apply them.  So having a smaller set of labels to learn is very helpful.

There will be project specific labels that only make sense to have in a specific repository, but there are significant number of core labels that apply to every repository of the organization.

I've been trying to come up with a standardized labeling in my own repositories. I use a `category: item` system. Even though this makes the labels pretty long, it's super nice to have all items for a given category sort together. Both the `arduino/Arduino` and `arduino/arduino-cli` repos are using a similar system to some extent.

This is the list I have so far for my own labels in case they might provide some inspiration:

- conclusion
  - declined
    - This will not be worked on
  - duplicate
    - This issue or pull request already exists
  - invalid
    - Off topic for this repository, or a bug report determined to not actually represent a bug 
- help wanted
  - Assistance from the community is especially welcome
  - Notes:
    - I would prefer put this in the status category, but GitHub has special handling of this label, so I think I should leave it just as `help wanted`, rather than `status: help wanted`.
- priority
  - high
  - low
  - Notes:
    - if no priority label is added, that is "medium"
- status
  - changes requested
  - in progress
  - on hold
  - pending
    - Waiting on action in another project
  - waiting for information
  - blocked
  - moved?
    - I'm ending up moving a lot of issues to Jira, and wishing I had a label to indicate this without needing to look at the issue. However, since GitHub has a transfer system for moving within my repos, I'm not sure how common this is. Probably better to just have it as a local label, or at least not a universal one.
- topic
  - CI
  - code
  - content
    - Notes:
      - Some repos don't contain code, so this could be the equivalent to "code" for those repos
        - example: ino-hardware-package-list, ino-troubleshooting
          - This might seem like "documentation", but I would like "documentation" to be about the documentation for the project, rather than about the content of a project when that content happens to be documentation.
  - documentation
  - metadata?
  - repository maintenance
    - I guess this would be about things like issues, PRs, contributing.md, labels, topics, repo description, wiki
  - structure
    - Notes:
      - this is very vague. I can't even really remember what I intended. Is it about the folder/filename structure of the repository? If so, it could be "repository structure"
- type
  - bug
  - enhancement
  - to-do
    - Notes:
      - I need to clearly define what this is. I thought it might be useful to have a "to-do list" item, but I haven't actually ended up using this much.

It's also important to make sure the labels are clearly documented and their use well defined. GitHub's label description feature can be used to add a short explanation of the meaning of each label.

One thing I found really useful for maintaining the `arduino/Arduino` issue tracker was when we established a policy (https://github.com/arduino/Arduino/issues/8642) of applying the "feature request" or "bug" labels according to to the intention of the user, rather than as a judgement call by the maintainer. I find that issues can almost always be put under one category or the other, and I can almost always filter my searches accordingly.

There is a GitHub Actions action that may be helpful in keeping the universal labels synced across all the repositories: https://github.com/lannonbr/issue-label-manager-action.

I have done some experimentation for my personal repositories with pulling a list of standardized universal labels from a centralized source, merging that with a list of repository-specific labels, and using the action to update the labels according to changes in the list:
```yaml
name: 'Manage repository labels'

on:
  push:
    paths:
      - '.github/workflows/manage-repository-labels.yml'
      - '.github/repository-labels-data/*.json'
  schedule:
    # run every Tuesday at 3 AM UTC
    - cron: '0 3 * * 2'
  workflow_dispatch:

jobs:
  manage-labels:
    runs-on: 'ubuntu-latest'

    env:
      LOCAL_REPOSITORY_LABELS_DATA_INPUTS_FOLDER: '.github/repository-labels-data'
      # Folder under runner.temp that will be used for merging the labels data files
      REPOSITORY_LABELS_DATA_MERGE_FOLDER: 'repository-labels-data'
      # See: https://github.com/lannonbr/issue-label-manager-action#usage
      REPOSITORY_LABELS_DATA_OUTPUT_PATH: '.github/labels.json'

    steps:
      - uses: 'actions/checkout@1.0.0'

      - name: 'Download universal labels file'
        uses: 'carlosperate/download-file-action@v1.0.3'
        with:
          
          file-url: 'https://raw.githubusercontent.com/per1234/GitHub-labels/main/universal-repository-labels.json'
          location: "${{ runner.temp }}/$REPOSITORY_LABELS_DATA_MERGE_FOLDER"

      - name: 'Merge labels data files'
        run: |
          REPOSITORY_LABELS_DATA_MERGE_PATH="${{ runner.temp }}/$REPOSITORY_LABELS_DATA_MERGE_FOLDER"
          cp "$LOCAL_REPOSITORY_LABELS_DATA_INPUTS_FOLDER"/*.json "$REPOSITORY_LABELS_DATA_MERGE_PATH"
          # Merge all data files and output to the final location
          jq -s '.=.|add|.' "$REPOSITORY_LABELS_DATA_MERGE_PATH"/*.json > "$REPOSITORY_LABELS_DATA_OUTPUT_PATH"

      - name: 'Manage labels'
        uses: 'lannonbr/issue-label-manager-action@2.0.0'
        env:
          GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
```

**Comment by aentinger** [comment]
Good morning @per1234 :wave: :tea: 

Thank you very much for picking up my comment in the other issue :+1: I'm glad to see that I'm not the only one who's a bit unhappy with the default labels provided by GitHub. That being said, I'm not on full agreement with all those labels you've enumerated (although I'm in full favour of the `category:item` system you propose).

Here's what I can imagine making sense for the 107-systems group of repositories:

* conclusion [declined | duplicate | invalid]
* help wanted
* priority [high | low]
* status [changes requested | in progress | on hold | waiting for information | blocked | moved]
I would eliminate `pending` as it's the same as `on hold` imho. `moved` also makes sense since we could choose to deal with a issue reported in one repository with a solution in another repository. Still I somehow feel there are still too many status messages.
* topic [ci, firmware, software, mechanics, electronics, documentation]
I'd tailor the labels towards the topics applicable to the Viper quadcopter project. That being said I'm not really happy with `mechanics` and `electronics`. Can you think of better words or is that the best we got? Maybe we should divide along Hardware:Mechanics|Sensors|Actuators ... .
* type [bug|feature request|support]
I'd add a `support` label since people usually also open issues when they've got a problem. What do you think? Also I'd rename `enhancement` to `feature request`.

@generationmake Please also have a look at this and share your thoughts.



PS: I've tried to access your github label definition file `https://raw.githubusercontent.com/per1234/GitHub-labels/main/universal-repository-labels.json` and could not obtain it. It looks like you've deleted that repository.

**Comment by per1234** [comment]
> Can you think of better words or is that the best we got?

- `topic: bits`
- `topic: atoms`

😄

>I'd add a support label since people usually also open issues when they've got a problem.

If support requests are something you want to support (haha) in the issue tracker, then I think that is a good idea. I don't allow this in my own repos, so I would just direct the issue author to an appropriate place to request support (e.g., Arduino Forum) and close the issue as invalid. But that's only my own approach.

>I'd rename enhancement to feature request.

The reason I chose "enhancement" instead of "feature request" is because I feel it applies better to pull requests. But it's a very minor distinction so I think either is fine.

>I've tried to access your github label definition file https://raw.githubusercontent.com/per1234/GitHub-labels/main/universal-repository-labels.json and could not obtain it. It looks like you've deleted that repository.

It was only a hello world for the sake of a PoC. I never got the time to actually fill in the real topics and put the system into action.

You can see the format of the JSON file in this screenshot from the action's readme:
![](https://github.com/lannonbr/issue-label-manager-action/blob/master/screenshots/json.png?raw=true)

**Comment by aentinger** [comment]
`topic: bits/atoms`: :rofl: That would be a most base-level distinction that certainly everyone could understand. Still I think I prefer the more concrete categories I outlined above :wink: 

Regarding "enhancement" vs "feature request" I'm with you that the former better resonates with pull requests while the latter better resonates with "feature demands" made in issues. Maybe we can have both?

The JSON file looks simple enough, I just wondered if you already had one in place for your own repositories.

**Comment by aentinger** [comment]
Hi @generationmake :wave: ☕ Can you please take a look at voice your opinion?

**Comment by aentinger** [comment]
Pinging @generationmake once more :point_right:

**Comment by per1234** [comment]
OK, I have a draft of the JSON file defining the universal labels for the 107-systems repositories:
```json
[
  {
    "name": "conclusion: declined",
    "color": "#b60205",
    "description": "Will not be worked on."
  },
  {
    "name": "conclusion: duplicate",
    "color": "#ff0000",
    "description": "Has already been submitted."
  },
  {
    "name": "conclusion: invalid",
    "color": "#ff0000",
    "description": "Off topic or not a bug."
  },
  {
    "name": "help wanted",
    "color": "#ffa200",
    "description": "Assistance from the community is especially welcome."
  },
  {
    "name": "priority: high",
    "color": "#ffa200",
    "description": ""
  },
  {
    "name": "priority: low",
    "color": "#c0c0c0",
    "description": ""
  },
  {
    "name": "status: changes requested",
    "color": "#ffa200",
    "description": "Changes to the PR are required before merging."
  },
  {
    "name": "status: in progress",
    "color": "#0000ff",
    "description": "Work is in progress on this."
  },
  {
    "name": "status: on hold",
    "color": "#b60205",
    "description": "Do not proceed at this time."
  },
  {
    "name": "status: waiting for information",
    "color": "#ffff00",
    "description": "More information must be provided before work can proceed."
  },
  {
    "name": "status: blocked",
    "color": "#b60205",
    "description": "Progress on this is prevented by an external cause."
  },
  {
    "name": "status: moved",
    "color": "#000080",
    "description": "This item was moved to another repository."
  },
  {
    "name": "topic: ci",
    "color": "#00ffff",
    "description": "Continuous integration for this repository."
  },
  {
    "name": "topic: firmware",
    "color": "#00ffff",
    "description": "Code that runs on an embedded system."
  },
  {
    "name": "topic: software",
    "color": "#00ffff",
    "description": "Code that runs on a PC."
  },
  {
    "name": "topic: mechanics",
    "color": "#00ffff",
    "description": "Mechanical components of the project."
  },
  {
    "name": "topic: electronics",
    "color": "#00ffff",
    "description": "Electronics components of the project."
  },
  {
    "name": "topic: documentation",
    "color": "#00ffff",
    "description": "Documentation for the project."
  },
  {
    "name": "type: bug",
    "color": "#ff0000",
    "description": ""
  },
  {
    "name": "type: enhancement",
    "color": "#008000",
    "description": "PR to improve the project."
  },
  {
    "name": "type: feature request",
    "color": "#008000",
    "description": "Issue requesting a new feature to be added."
  },
  {
    "name": "type: support",
    "color": "#ffff00",
    "description": "Issue requesting a new feature to be added."
  }
]
```
Here you can see the colors:
![Clipboard01](https://user-images.githubusercontent.com/8572152/100571487-18538580-3288-11eb-8d22-b039d2922ef9.png)

I welcome any feedback.

Once this is finalized, the next question is where the universal labels definition file should be published. I was thinking the `107-systems/.github` repository could be a good location, as this is a standard location for storing files used by multiple repositories. GitHub has some special features based around it:
https://docs.github.com/en/free-pro-team@latest/github/building-a-strong-community/creating-a-default-community-health-file#creating-a-repository-for-default-files
so you might end up creating this repository eventually anyway. But to be clear, there is no special significance to the `107-systems/.github` repository for this particular usage.

**Comment by aentinger** [comment]
That's fantastic! Great work @per1234 :rocket: I'm sorry that @generationmake didn't come back with feedback but let's move ahead anyway ... I didn't know about this great feature of default files - thank you for bringing it to my attention. I've created one for 107-Systems [here](https://github.com/107-systems/.github). Can I please ask you to move ahead and integrate your JSON there as well as setup the action to import that changes into at least this repository?

**Comment by per1234** [comment]
OK, PR submitted: https://github.com/107-systems/.github/pull/1
I'll submit a follow-up PR adding the CI workflow to this repo once that is in place.

>I didn't know about this great feature of default files - thank you for bringing it to my attention.

There is another feature not mentioned at the link I posted in my previous reply. This is the ability to create custom GitHub Actions workflow "templates" (I actually think "stock" or "model" would be a better term than "template", since they don't have any templating functionality) that can be added to any repository in the organization with just a couple clicks in the GitHub web interface:
https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/sharing-workflows-with-your-organization#creating-a-workflow-template

**Comment by aentinger** [comment]
Thank you very much, I've merged https://github.com/107-systems/.github/pull/1 in. Please proceed with the action for accessing that workflow. I didn't know about this workflow template capabilities either but it looks to me like it could solve a lot of work for you in the future, because we can self-service setup CI build for future libraries via a couple of clicks, amirite?

**Comment by generationmake** [comment]
Hi.
Sorry for not responding earlier.
Until now I haven't used the labels very much so I think I cannot contribute much to this conversation. 
All your proposals make sense and I think we should just give it a try.

**Comment by aentinger** [comment]
I've just used them a bit, but I used them enough to get the feeling that the standard labels aren't good enough. For the future I'd like to ask you to try using them. I'm closing this issue since it has been dealt with by #29.

**Comment by per1234** [comment]
>we can self-service setup CI build for future libraries via a couple of clicks, amirite?

For the most part. It might sometimes be necessary to make some adjustments. The Compile Examples workflow tends to be different from one repo to another, but actually I don't think there is that much variation so far in the 107-systems workflows.

The other workflows would only take some small adjustments.

I like the idea of having a centralized location to pull workflows from. Currently, I try to remember which repo has the most "state of the art" workflow, then I do a copy/paste of that one and make the necessary modifications, but I have trouble remembering which is the best one.

>All your proposals make sense and I think we should just give it a try.

This is a really great thing about this system. If we later find the need for additional labels, they only need to be added to a single repository, then they automatically propagate across the organization from there.

I think it's also a lot easier to use labels that are consistent. Currently, I'm maintaining repositories with something like 5 completely distinct sets of labels. I have trouble remembering which labels to search for and there are labels that I don't fully understand the meaning of, and so am hesitant to use.
